rfcs/53-consistent-hierarchy.md

```diff
@@ -0,0 +1,421 @@
+parent-sibling-child hierarchies currently found in `bevy_transform`. This should
+also allow us to internally leverage the same decoupling too for specialized
+transforms for 2D and UI.
+
+## Benefits
+
+ - This rocks the boat the least compared to the following alternatives and
+   largely provides the same benefits.
+ - This design has (nearly) guarenteed global consistency. The hierarchy cannot
+   be arbitrarily mutated in inconsistent between synchronization points.
+   Changes are deferred to a until the end of the current stage instead of a
+   single global maintainence system.
+ - The cache locality when iterating over the immediate children of an entity is
+   retained. This saves an extra cache miss per entity when traversing the
+   hierarchy.
+ - The removal of `PreviousParent` should reduce the archetype fragmentation of
+   the current solution.
+ - Any update to the hierarchy is "atomic" and is `O(n)` in the number of
+   children an entity has. For any practical hierarchy, this will typically be
+   a very small n.
+ - Existing query filters still work as is. For example, `Without<Parent>` can
+   still be used to find root entities, and `Changed<Children>` can be used to
+   find any entity with either new children, removed children, or reordered
+   children.
+ - The hierarchy is decoupled from the base transform system, allowing both Bevy
+   developer and third-party crates to make their own independent transform
+   system without compiling the default transform system.
+
+## Drawbacks
+
+ - The fact that the hierarchy is not optimally stored in memory is still an
+   issue, and all hiearchy traversals require heavy random access into memory.
+ - Updates are still not immediately visible within a stage and deferred until
+   the next command buffer flush.
+ - Hierarchy updates are now single threaded. Commands can be generated from
+   multiple systems at the same time, but they now require exclusive access to
+   apply.
+ - Each of the changed commands do a lot more work before and are signifgantly
+   more branch heavy. This could negatively impact the performance of
+   command-heavy workloads (i.e.  scene loading)
+ - The commands applied for controlling the hierarchy are computationally more
+   expensive, requiring more branching access and more `World::get` calls.
+ - Some potential implementations of this hierarchy requires normal ECS commands
+   to be aware of the hierarchy. Using `EntityCommands::remove` to remove
+   `Children` or `Parent` will break the invariants of the system without some
+   hooks to enforce the invariants on removal. This is partially mitigatable via
+   documentation until hooks or relations lands.
+
+## Rationale and alternatives
+The primary goal of this proposal is to get rid of the hierarchy maintenance
+system and the consistency issues that come with it's design, which it delivers
+on.
+
+### Alternative: Hierarchy as a Resource
```

Comment:
The biggest downside here is definitely scene management. We'd need to either:

1. Add the concept of "resource merging". In practice I think this could be reasonably cheap, especially if we assume that scenes have a single root entity (and I'm relatively convinced at this point that they should).
2. Add a secondary component-based hierarchy representation for scenes, which then gets "converted" into a runtime centralized Hierarchy resource when the scene is spawned.

Both of these are fundamental changes to the mental model that require serious thought.

> Iteration requires mutative access to the resource, which can limit parallelism of systems traversing the hierarchy. The alternative requires a dedicated O(n) sweep over the hierarchy whenever it's mutated.

We don't necessarily need to couple these operations. I get that we're deferring topological sorts to avoid really expensive insertion (we don't want to be constantly shifting items around on each insert). But this is an optimization issue, not a correctness issue. We could choose to do it once per frame with `traverse_hierarchy_and_toposort(&mut self)` (maybe in the transform propagation) and encourage users to just do `traverse_hierarchy(&self)` unless they really want to toposort earlier. On average, iterations would be approximately topo-sorted. Or we could have a "dirty" bit and just topo sort on frames where it is dirty.

Reply:
Updated the RFC to reflect some of these + some pros/cons I initially missed. I actually like this a lot more now that it's clear that this even further addresses the archetype fragmentation and makes hierarchy edits on par with a SparseSet lookup.

If we know the roots of a scene, assuming we don't need to ensure it's sorted immediately afterwards, copy-pasting the mapped contents of a scene hierarchy at the end of the "global hierarchy" then updating the spawned scene roots to their new parent(s) should be `O(n)` in the number of entities in the hierarchy, and should be very fast. It can then rely on the frame-to-frame topological sort to optimize out the kinks.